src/ApiService/ApiService/OneFuzzTypes/Metrics.cs

```diff
@@ -0,0 +1,376 @@
+    NodeStateUpdated,
+    Ping,
+    PoolCreated,
+    PoolDeleted,
+    ProxyCreated,
+    ProxyDeleted,
+    ProxyFailed,
+    ProxyStateUpdated,
+    ScalesetCreated,
+    ScalesetDeleted,
+    ScalesetFailed,
+    ScalesetStateUpdated,
+    ScalesetResizeScheduled,
+    TaskCreated,
+    TaskFailed,
+    TaskStateUpdated,
+    TaskStopped,
+    CrashReported,
+    RegressionReported,
+    FileAdded,
+    TaskHeartbeat,
+    NodeHeartbeat,
+    InstanceConfigUpdated,
+    NotificationFailed
+}
+
+public abstract record BaseMetric() {
+
+    private static readonly IReadOnlyDictionary<Type, MetricType> typeToMetric;
+    private static readonly IReadOnlyDictionary<MetricType, Type> metricToType;
+    private static int metricValue;
+
+    static BaseMetric() {
+
+        MetricType ExtractMetricType(Type type) {
+            var attr = type.GetCustomAttribute<MetricTypeAttribute>();
+            if (attr is null) {
+                throw new InvalidOperationException($"Type {type} is missing {nameof(MetricTypeAttribute)}");
+            }
+            return attr.MetricType;
+        }
+
+        typeToMetric =
+            typeof(BaseMetric).Assembly.GetTypes()
+            .Where(t => t.IsSubclassOf(typeof(BaseMetric)))
+            .ToDictionary(x => x, ExtractMetricType);
+
+        metricToType = typeToMetric.ToDictionary(x => x.Value, x => x.Key);
+
+        // check that all Metric types are accounted for
+        var allMetricTypes = Enum.GetValues<MetricType>();
+        var missingMetricTypes = allMetricTypes.Except(metricToType.Keys).ToList();
+        if (missingMetricTypes.Any()) {
+            throw new InvalidOperationException($"Missing metric types: {string.Join(", ", missingMetricTypes)}");
+        }
+
+        metricValue = 1;
+    }
+
+
+    public MetricType GetMetricType() {
+        var type = this.GetType();
+        if (typeToMetric.TryGetValue(type, out var metricType)) {
+            return metricType;
+        }
+
+        throw new NotSupportedException($"Unknown metric type: {type.GetType()}");
+    }
+
+    public static Type GetTypeInfo(MetricType metricType) {
+        if (metricToType.TryGetValue(metricType, out var type)) {
+            return type;
+        }
+
+        throw new ArgumentException($"Unknown metric type: {metricType}");
+    }
+
+    public static void SetMetricValue(int value) {
+        metricValue = value;
+    }
+
+    public static int GetMetricValue() {
+        return metricValue;
+    }
+};
+
+public class MetricTypeProvider : ITypeProvider {
+    public Type GetTypeInfo(object input) {
+        return BaseMetric.GetTypeInfo((input as MetricType?) ?? throw new ArgumentException($"input is expected to be an MetricType {input}"));
+    }
+}
+
+[MetricType(MetricType.TaskStopped)]
+public record MetricTaskStopped(
+    Guid JobId,
+    Guid TaskId,
+    UserInfo? UserInfo,
+    TaskConfig Config
+) : BaseMetric();
+
+[MetricType(MetricType.TaskFailed)]
+public record MetricTaskFailed(
```

Comment:
These definitions look like the Events. Do we need new records for those? 
We could just use the event and derive the Metric form them

Reply:
They are essentially duplicates right now, but that was just meant to be a starting point. I'm thinking that the actual fields may change as we go along to include different info. For example, I updated the TaskHeartbeat _metric_ record to be different from the event one.

Reply:
For ease, I just duplicated all existing events with the idea being that many of the events will be metrics at some point.